### PR TITLE
[perms] Implement getTeams WEB-501

### PR DESCRIPTION
**Pull request by easyCZ**
## Related Issue(s)

Fixes #

#### Preview status

gitpod:summary

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [ ] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`
</details>

/hold

### CODE REVIEW

File: components/server/src/workspace/gitpod-server-impl.ts [hunk 1]
```diff
@@ -2734,7 +2734,29 @@ export class GitpodServerImpl implements GitpodServerWithTracing, Disposable {
     public async getTeams(ctx: TraceContext): Promise<Team[]> {
         // Note: this operation is per-user only, hence needs no resource guard
         const user = await this.checkUser("getTeams");
-        return this.teamDB.findTeamsByUser(user.id);
+        const teams = await this.teamDB.findTeamsByUser(user.id);
+
+        // We need to check each team individually against our permission system.
+        // checks are promises, which resolve to { team, check } object
+        const checks = teams.map((team) =>
+            this.authorizer.check(ReadOrganizationInfo(user.id, team.id)).then((check) => ({ team, check })),
```
Comment (easyCZ) [inline components/server/src/workspace/gitpod-server-impl.ts:2742]: bundling `team` and `check` together avoids having to create a lookup map of teams by ID

File: components/server/src/workspace/gitpod-server-impl.ts [hunk 2]
```diff
@@ -2734,7 +2734,29 @@ export class GitpodServerImpl implements GitpodServerWithTracing, Disposable {
...
+        );
+        const checkResults = await Promise.allSettled(checks);
+
+        const accessibleTeams = [];
+        for (let result of checkResults) {
+            if (result.status !== "fulfilled") {
```
Comment (svenefftinge) [inline components/server/src/workspace/gitpod-server-impl.ts:2748]: we should log an error if the promise was rejected

**Review by svenefftinge — APPROVED** [review]
lgtm minus the comment

File: components/server/src/workspace/gitpod-server-impl.ts (hunk 2 above)
Comment (easyCZ) [reply components/server/src/workspace/gitpod-server-impl.ts:2748]: Sure, can do. What do we gain from this?

I'll first collect all rejected and log only once after the loop, if that's fine by you

Comment (svenefftinge) [reply components/server/src/workspace/gitpod-server-impl.ts:2748]: Would be good to log errors coming from `check` as we would optherwise silently swallow them and treat them as not authorized.
You can alternatively add a catch clause after the then in line 2742.

Comment (easyCZ) [reply components/server/src/workspace/gitpod-server-impl.ts:2748]: > Would be good to log errors coming from check 
we already do log these on the `authorizer` level, but not necessarily in the context of the RPC (but the operation is included in the log)

Comment (svenefftinge) [reply components/server/src/workspace/gitpod-server-impl.ts:2748]: Ok, I couldn't tell looking at the code. In general, code that swallows possible exceptions is better to be avoided. We have linting for this but, of course, it is not intelligent enough to check all `allSettled` states.

Comment (easyCZ) [reply components/server/src/workspace/gitpod-server-impl.ts:2748]: Added logging for the rejected case

**Comment by easyCZ** [comment]
/unhold

Tested on this preview https://mp-perms-get-teams-2.preview.gitpod-dev.com/workspaces as in this PR, I couldn't get it to deploy

**Pull request merged**